utils/sipquery.1
.\" sipquery - register and report sip contact details
.\" Copyright (c) 2009-2014 David Sugar <dyfet@gnutelephony.org>
.\" Copyright (c) 2015-2017 Cherokees of Idaho.
.\"
.\" This manual page is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU Lesser General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.\" This manual page is written especially for Debian GNU/Linux.
.\"
.TH sipquery "1" "December 2009" "GNU SIP Witch" "GNU Telephony"
.SH NAME
sipquery \- register and report sip contact details
.SH SYNOPSIS
.B sipquery
.RB [ options ]
.I userid[@server[:port]]
.SH DESCRIPTION
This is a tool that uses the registration process to query the sip contact
uri for a registered user agent.  This may even be considered a demonstration
of a security flaw in the standard SIP server model.  Basically, an
un-authenticated registration request is sent to the server, and the contact
uri is read to standard out from the server reply message.
.SH OPTIONS
.TP
.BI \-\-forward= address
Sets the "public appearing" IP address for testing behind NAT.
.TP
.BI \-\-port= number
Sets the sip port number, in case the default port (5060) is in use already.
.TP
.BI \-\-proxy= host
Use a specified outbound proxy uri to get to the server being queried.
.TP
.BI \-\-server= host
Specify server uri to query unless offered as part of userid.
.TP
.BI \-\-timeout= seconds
Time to wait for query response.
.SH AUTHOR
.B sipquery
was written by David Sugar <dyfet@gnutelephony.org>.
.SH "REPORTING BUGS"
Report bugs to sipwitch-devel@gnu.org or bugs@gnutelephony.org.
.SH COPYRIGHT
Copyright \(co 2009-2014 David Sugar, Tycho Softworks.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.



NEWS
SIP Witch NEWS -- history of visible user changes. Updated 2008-04-15

SIP Witch is an official package of the GNU Project as of August 10th 2007.
GNU SIP Witch is also part of GNU Telephony & the GNU Telecom subsystem.

sipwitch 0.1 codebase initial public release with support of SIP registration,
multi-target registration, authentication peering over SIP, and registration
querying of user agents, basic call processing between registered user agents
including call distribution for multi-target registrations, and basic SIP
instant messaging support.

sipwitch 0.4 represents the first minimally deployable baseline sipwitch
server.  This release includes multipath calling, and basic call forwarding.
Also important with this release is stablization of core management api's,
including SWIG interfaces for local scripting languages, and xmlrpc cgi.
Basic management and administration services can now be built for sipwitch.
Support for changing server state is also complete, such as for "day"/"night"
modes, holidays, etc.  Basic server call statistics and call detail record
processing is completed.

sipwitch 0.5 introduces the concept of internodal calling and allows sipwitch
to transform itself into a sip redirection server for certain local call
destinations.  Internodal calling differs from dialing to remote uri's in that
the authentication record is pre-stuffed into the remote invite, hence, all
nodes must share the same sip realm.  This allows one sipwitch server to call a
"local" extension number that actually appears on a remote switch rather than
the one the user agent is registered to. Internodal references are generated
using pattern rules through the static <routing> table <redirect> entries.  The
forward secure calling domain plugin also makes use of internodal calling to
connect with the insecure domain.
  


utils/sipquery.cpp
// Copyright (C) 2008-2014 David Sugar, Tycho Softworks.
// Copyright (C) 2015-2017 Cherokees of Idaho.
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

#undef  HAVE_CONFIG_H

#include <sipwitch-config.h>
#include <eXosip2/eXosip.h>
#include <ucommon/ucommon.h>

static int verbose = 0;
static int port = 0;
static int family = AF_INET;
static int protocol = IPPROTO_UDP;
static const char *server = NULL;
static const char *forwarded = NULL;
static const char *proxy = NULL;
static const char *binding = NULL;
static unsigned timeout = 1000;
static int tls = 0;

#if defined(EXOSIP_OPT_BASE_OPTION) && !defined(EXOSIP_OPT_DONT_SEND_101)
#define EXOSIP_API4
#endif

#ifdef  EXOSIP_API4
#define EXOSIP_CONTEXT  context
#define OPTION_CONTEXT  context,
#define EXOSIP_LOCK     eXosip_lock(context);
#define EXOSIP_UNLOCK   eXosip_unlock(context);
static struct eXosip_t *context = NULL;
#else
#define EXOSIP_LOCK     eXosip_lock();
#define EXOSIP_UNLOCK   eXosip_unlock();
#define EXOSIP_CONTEXT
#define OPTION_CONTEXT
#endif

using namespace ucommon;

#if defined(_MSWINDOWS_) && defined(__GNUC__)
// binds addrinfo for mingw32 linkage since otherwise mingw32 cannot
// cannot link proper getaddrinfo/freeaddrinfo calls that eXosip uses.
static Socket::address localhost("127.0.0.1");
#endif

PROGRAM_MAIN(argc, argv)
{
    int error = 2;
    eXosip_event_t *sevent;
    const char *cp, *user;
    char buffer[256];
    char pbuffer[256];
    char tbuffer[256];
    osip_message_t *msg = NULL;
    int rid;
    osip_contact_t *contact;
    int pos = 0;
    int stop = 0;

    cp = getenv("SIP_PROXY");
    if(cp)
        proxy = cp;

    cp = getenv("SIP_SERVER");
    if(cp)
        server = cp;


    while(NULL != *(++argv)) {
        if(!strcmp(*argv, "--")) {
            ++argv;
            break;
        }

        if(!strncmp(*argv, "--", 2))
            ++*argv;

        if(!strcmp(*argv, "-q") || !strcmp(*argv, "-quiet")) {
            verbose = 0;
            continue;
        }

        if(!strcmp(*argv, "-version")) {
            printf("sipquery 0.1\n"
                "Copyright (C) 2008 David Sugar, Tycho Softworks\n"
                "License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>\n"
                "This is free software: you are free to change and redistribute it.\n"
                "There is NO WARRANTY, to the extent permitted by law.\n");
            exit(0);
        }

        if(!strcmp(*argv, "-v") || !strcmp(*argv, "-verbose")) {
            verbose = 1;
            continue;
        }


        if(!strcmp(*argv, "-t") || !strcmp(*argv, "-timeout")) {
            if(NULL == *(++argv))
                shell::errexit(3, "*** sipuser: timeout option missing timeout\n");
            timeout = atoi(*argv) * 1000;
            if(!timeout)
                shell::errexit(3, "*** sipuser: timeout option invalid\n");

            continue;
        }

        if(!strncmp(*argv, "-timeout=", 9)) {
            timeout = atoi(*argv + 9) * 1000;
            if(!timeout)
                shell::errexit(3, "*** sipuser: timeout option invalid\n");

            continue;
        }

        if(!strcmp(*argv, "-p") || !strcmp(*argv, "-port")) {
            if(NULL == *(++argv))
                shell::errexit(3, "*** sipuser: port option missing port\n");

            port = atoi(*argv);
            if(!port)
                shell::errexit(3, "*** sipuser: port option invalid number\n");

            continue;
        }

        if(!strncmp(*argv, "-port=", 6)) {
            port = atoi(*argv + 6);
             if(!port)
                shell::errexit(3, "*** sipuser: port option invalid number\n");

            continue;
        }

        if(!strcmp(*argv, "-proxy")) {
            proxy = *(++argv);
            if(!proxy)
                shell::errexit(3, "*** sipuser: proxy option missing proxy\n");
            continue;
        }

        if(!strncmp(*argv, "-proxy=", 7)) {
            proxy = *argv + 7;
            continue;
        }

        if(!strcmp(*argv, "-server")) {
            server = *(++argv);
            if(!server)
                shell::errexit(3, "*** sipuser: server option missing proxy\n");
            continue;
        }

        if(!strncmp(*argv, "-server=", 8)) {
            server = *argv + 8;
            continue;
        }

        if(!strcmp(*argv, "-forward")) {
            forwarded = *(++argv);
            if(!forwarded)
                shell::errexit(3, "*** sipuser: forwarding option missing interface\n");
            continue;
        }

        if(!strncmp(*argv, "-forward=", 9)) {
            forwarded = *argv + 8;
            continue;
        }

        if(!strcmp(*argv, "-?") || !strcmp(*argv, "-h") || !strcmp(*argv, "-help")) {
            fprintf(stderr, "usage: sipuser [options] userid\n"
                "[options]\n"
                "  -proxy    sip:proxyhost[:port]\n"
                "  -server   sip:server[:port]\n"
                "  -forward  ip-address\n"
                "  -port     port-numer\n"
                "  -timeout  seconds\n"
                "  -verbsose\n"
                "Report bugs to sipwitch-devel@gnu.org\n");
            exit(3);
        }

        if(**argv == '-')
            shell::errexit(3, "*** sipuser: %s: unknown option\n", *argv);

        break;
    }

    if(!*argv) {
usage:
        shell::errexit(3, "use: sipuser [options] userid\n");
    }

    user = *(argv++);
    if(*argv)
        goto usage;

#if defined(WIN32) || defined(_WIn32)
    if(!port)
        port = 5060;
#else
    if(!port)
        port = 5060 + getuid();
#endif

#ifdef  EXOSIP_API4
    context = eXosip_malloc();
#endif

    if(eXosip_init(EXOSIP_CONTEXT))
        shell::errexit(3, "*** sipuser: failed exosip init\n");

#ifdef  AF_INET6
    if(family == AF_INET6) {
        eXosip_enable_ipv6(1);
        if(binding == NULL)
            binding = "::0";
    }
#endif

    if(eXosip_listen_addr(OPTION_CONTEXT protocol, binding, port, family, tls)) {
#ifdef  AF_INET6
        if(!binding && family == AF_INET6)
            binding = "::0";
#endif
        if(!binding)
            binding = "*";
        shell::errexit(3, "*** sipuser: failed to listen %s:%d\n", binding, port);
    }

    if(forwarded)
        eXosip_masquerade_contact(OPTION_CONTEXT forwarded, port);

    eXosip_set_user_agent(OPTION_CONTEXT "SIPW/sipquery");

    if(!strncmp(user, "sip:", 4)) {
        tls = 0;
        user += 4;
    }
    else if(!strncmp(user, "sips:", 5)) {
        tls = 1;
        user += 5;
    }

    if(!server && strchr(user, '@')) {
        server = strchr(user, '@');
        ++server;
    }

    if(server && !strncmp(server, "sip:", 4))
        server += 4;
    else if(server && !strncmp(server, "sips:", 5))
        server += 5;

#ifdef  AF_INET6
    if(family == AF_INET6 && server == NULL)
        server = "::1";
#endif
    if(server == NULL)
        server = "127.0.0.1";

    if(!proxy)
        proxy = server;

    if(strncmp(proxy, "sip:", 4) && strncmp(proxy, "sips:", 5)) {
        if(tls)
            snprintf(pbuffer, sizeof(pbuffer), "sips:%s", proxy);
        else
            snprintf(pbuffer, sizeof(pbuffer), "sip:%s", proxy);
        proxy = pbuffer;
    }

    if(tls && !strchr(user, '@'))
        snprintf(buffer, sizeof(buffer), "sips:%s@%s", user, server);
    else if(!strchr(user, '@'))
        snprintf(buffer, sizeof(buffer), "sip:%s@%s", user, server);
    else if(tls)
        snprintf(buffer, sizeof(buffer), "sips:%s", user);
    else
        snprintf(buffer, sizeof(buffer), "sip:%s", user);


    EXOSIP_LOCK
    rid = eXosip_register_build_initial_register(OPTION_CONTEXT buffer, proxy, NULL, 60, &msg);
    if(!msg) {
        EXOSIP_UNLOCK
        shell::errexit(3, "*** sipuser: cannot create query for %s\n", user);
    }
    snprintf(tbuffer, sizeof(tbuffer), "<%s>", buffer);
    osip_message_set_to(msg, tbuffer);
    osip_list_ofchar_free(&msg->contacts);
    eXosip_register_send_register(OPTION_CONTEXT rid, msg);
    EXOSIP_UNLOCK

    while(!stop) {
        sevent = eXosip_event_wait(OPTION_CONTEXT 0, timeout);
        if(!sevent)
            shell::errexit(2, "*** sipuser: timed out\n");

        if(sevent->type == EXOSIP_REGISTRATION_FAILURE) {
            error = 1;
            ++stop;
        }

        if(sevent->type == EXOSIP_REGISTRATION_SUCCESS) {
            pos = 0;
            while(verbose && sevent->response && !osip_list_eol(OSIP2_LIST_PTR sevent->response->contacts, pos)) {
                contact = (osip_contact_t *)osip_list_get(OSIP2_LIST_PTR sevent->response->contacts, pos++);
                if(contact && contact->url)
                    printf("%s:%s@%s:%s\n",
                        contact->url->scheme, contact->url->username,
                        contact->url->host, contact->url->port);
            }
            error = 0;
            ++stop;
        }

        eXosip_event_free(sevent);
    }
    eXosip_register_remove(OPTION_CONTEXT rid);
    eXosip_quit(EXOSIP_CONTEXT);
    PROGRAM_EXIT(error);
}

